src/Lecture2.hs

```diff
@@ -159,8 +171,30 @@ data Knight = Knight
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
-
-dragonFight = error "TODO"
+data Dragon a = Dragon
+  { dragonHealth    :: Int
+  , dragonFirePower :: Int
+  , dragonChest     :: Chest a}
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There're two problems with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
    { dragonHealth = 150
    , dragonFirePower = 150
    , dragonTreasure = Chest 10 (Just True)
    }
```

2. It's possible to not have a treasure inside non-greed dragons:

```haskell
myDragon = Dragon
    { dragonHealth = 150
    , dragonFirePower = 150
    , dragonTreasure = Chest 10 Nothing
    }
```

Reply:
Do you have any tips for this?

Reply:
You can define your dragon type as the following sum type to model the data type precisely:

```haskell
data Dragon a
    = Green DragonProperties Int
    | Red   DragonProperties (Chest a)
    | Black DragonProperties (Chest a)
```